packages/calling-stateful-client/src/CallAgentDeclarativeCommon.ts

```diff
@@ -0,0 +1,248 @@
+        incomingCall.id,
+        new IncomingCallSubscriber(incomingCall, this.setIncomingCallEnded)
+      );
+    }
+    this._declarativeIncomingCalls.set(incomingCall.id, incomingCallDeclaratify(incomingCall, this._context));
+    this._context.setIncomingCall(convertSdkIncomingCallToDeclarativeIncomingCall(incomingCall));
+  };
+
+  protected addCall = (call: CallCommon): DeclarativeCallCommon => {
+    this._callSubscribers.get(call)?.unsubscribe();
+
+    // For API extentions we need to have the call in the state when we are subscribing as we may want to update the
+    // state during the subscription process in the subscriber so we add the call to state before subscribing.
+    this._context.setCall(convertSdkCallToDeclarativeCall(call));
+    this._callSubscribers.set(call, new CallSubscriber(call, this._context, this._internalContext));
+    return this.getOrCreateDeclarativeCall(call);
+  };
+
+  private getOrCreateDeclarativeCall = (call: CallCommon): DeclarativeCallCommon => {
+    const declarativeCall = this._declarativeCalls.get(call);
+    if (declarativeCall) {
+      return declarativeCall;
+    }
+
+    const newDeclarativeCall = this.callDeclaratify(call, this._context);
+    this._declarativeCalls.set(call, newDeclarativeCall);
+    return newDeclarativeCall;
+  };
+
+  // args could be either from teamsCall or Call, set it as unknown and cast it in child class later
+  protected abstract startCall(agent: CallAgentCommon, args: unknown[]): CallCommon;
+  protected abstract joinCall(agent: CallAgentCommon, args: unknown[]): CallCommon;
+  protected abstract agentSubscribe(agent: CallAgentCommon, args: unknown[]): void;
+  protected abstract agentUnsubscribe(agent: CallAgentCommon, args: unknown[]): void;
+
+  // We can't directly override get function because it is proxied,
+  // Add a getCommon function and call it in child class
+  protected getCommon<
+    AgentType extends CallAgent | /* @conditional-compile-remove(teams-identity-support) */ TeamsCallAgent,
+    P extends keyof CallAgent
+  >(target: AgentType, prop: P): any {
+    switch (prop) {
+      case 'startCall': {
+        return this._context.withErrorTeedToState((...args: Parameters<AgentType['startCall']>): CallCommon => {
+          const call = this.startCall(target, args);
+          this.addCall(call);
+          return this.getOrCreateDeclarativeCall(call);
+        }, 'CallAgent.startCall');
+      }
+      case 'join': {
+        return this._context.withErrorTeedToState((...args: Parameters<AgentType['join']>): CallCommon => {
+          const call = this.joinCall(target, args);
+          this.addCall(call);
+          return this.getOrCreateDeclarativeCall(call);
+        }, 'CallAgent.join');
+      }
+      case 'calls': {
+        return Array.from(this._declarativeCalls.values());
+      }
+      case 'on': {
+        return (...args: Parameters<CallAgent['on']> | Parameters<AgentType['on']>): void => {
```

Comment:
This is the only function with `Parameters<CallAgent['PROP']>` (the rest only have `Parameters<AgentType['PROP']>` - can this be:
```suggestion
        return (...args: Parameters<AgentType['on']>): void => {
```